### ISSUE TITLE
Array literal type not fully inferred; blatant type error ignored

### ISSUE THREAD

**Issue by cpcallen** [google/closure-compiler#2896]
I was surprised to discover that OTI doesn't seem to catch the blatant type error in the following code:

```JS
/** @param {!Array<string>} foo */
var f = function(foo) {};

f([42]);
```

NTI catches this correctly.  The problem appears to be that the [compiler believes the type of `[42]` to be `Array`, and does not complain about using an `Array` as an `Array<string>`](https://closure-compiler-debugger.appspot.com/#input0%3D%252F**%2520%2540param%2520%257B!Array%253Cstring%253E%257D%2520foo%2520*%252F%250Avar%2520f%2520%253D%2520function(foo)%2520%257B%257D%253B%250A%250Af(%255B42%255D)%253B%250A%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3Dtrue%26CHECK_LINT%3Dtrue%26CHECK_SYMBOLS%3Dtrue%26CHECK_MISSING_RETURN%3Dtrue%26CHECK_SUSPICIOUS_CODE%3Dtrue%26MISSING_PROPERTIES%3Dtrue%26CHECK_REQUIRES%3Dtrue%26CHECK_PROVIDES%3Dtrue%26CHECK_GLOBAL_NAMES%3Dtrue%26CHECK_DEPRECATED%3Dtrue%26CHECK_VISIBILITY%3Dtrue%26CHECK_CONSTANTS%3Dtrue%26CHECK_ES5_STRICT%3Dtrue%26CHECK_GLOBAL_THIS%3Dtrue%26CHECK_UNREACHABLE_CODE%3Dtrue%26TRANSPILE%3Dtrue%26CHECK_TYPES%3Dtrue%26STRICT_CHECK_TYPES%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue).  (The same error will (fail to) occur [when assigning instead of passing as a parameter](https://closure-compiler-debugger.appspot.com/#input0%3D%252F**%2520%2540type%2520%257B!Array%253Cstring%253E%257D%2520*%252F%250Avar%2520a%253B%250A%250Aa%2520%253D%2520%255B42%255D%253B%250A%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3Dtrue%26CHECK_LINT%3Dtrue%26CHECK_SYMBOLS%3Dtrue%26CHECK_MISSING_RETURN%3Dtrue%26CHECK_SUSPICIOUS_CODE%3Dtrue%26MISSING_PROPERTIES%3Dtrue%26CHECK_REQUIRES%3Dtrue%26CHECK_PROVIDES%3Dtrue%26CHECK_GLOBAL_NAMES%3Dtrue%26CHECK_DEPRECATED%3Dtrue%26CHECK_VISIBILITY%3Dtrue%26CHECK_CONSTANTS%3Dtrue%26CHECK_ES5_STRICT%3Dtrue%26CHECK_GLOBAL_THIS%3Dtrue%26CHECK_UNREACHABLE_CODE%3Dtrue%26TRANSPILE%3Dtrue%26CHECK_TYPES%3Dtrue%26STRICT_CHECK_TYPES%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue).)

**Comment by MatrixFrog** [comment]
Yup. I think this is https://github.com/google/closure-compiler/issues/312

**Comment by cpcallen** [comment]
You are correct.  My search for previous reports was evidently not as thorough as I thought.
